R/benchTradePerf.R

```diff
@@ -0,0 +1,214 @@
+#'      \item{\code{Quality}: }{A qualitiative RPM score over quintiles, bottom-up one of 'Poor', 'Fair', 'Average', 'Good', 'Excellent'. Present if \code{verbose = TRUE}}
+#' }
+#' 
+#' @seealso \code{\link{initPortf}}, \code{\link{addTxn}}
+#' 
+#' @examples 
+#' 
+#' @references Kissell, R. \emph{The Science of Algorithmic Trading and Portfolio Management} (ISBN 978-0-12-401689-7)
+#' 
+#' @author Vito Lestingi
+#' 
+#' @export
+#' 
+benchTradePerf <- function(Portfolio,
+                           Symbol,
+                           side = 1,
+                           benchmark = c("MktBench", "VWAP", "PWP", "RPM"),
+                           type = c("Txns", "Mkt"),
+                           MktData,
+                           POV = NULL,
+                           verbose = FALSE
+                           )
+{ 
+  pname <- Portfolio
+  Portfolio <- .getPortfolio(pname)
+  txns <- Portfolio[["symbols"]][[Symbol]][["txn"]]
+  p_avg <- as.numeric(last(txns$Pos.Avg.Cost))
+  tTxnQty <- sum(txns$Txn.Qty)
+  
+  # Benchmark metrics
+  benchmark <- match.arg(benchmark, c("MktBench", "VWAP", "PWP", "RPM"), several.ok = TRUE)
+  
+  tradesPerf <- list()
+  
+  for (i in 1:length(benchmark)) {
+    switch (benchmark[i],
+            MktBench = {
+              if (!("MktPrice" %in% colnames(MktData))) stop("No MktPrice column found, what did you call it?")
+              mktBenchPrice <- as.numeric(MktData[1, "MktPrice"])
+              benchPrice <- mktBenchPrice
+              
+              out <- as.data.frame(cbind(Symbol, c("Buy", "Sell")[side], p_avg, benchPrice), stringsAsFactors = FALSE)
+              colnames(out) <- c('Symbol', 'Side', 'Avg.Exec.Price', paste(benchmark[i], type, sep = '.'))
+            },
+            VWAP = {# full VWAP
+              if (missing(type)) type <- 'Txns'
+              
+              if (type == 'Txns') {
+                vwapPrice <- crossprod(txns$Txn.Price, txns$Txn.Qty)/tTxnQty
+              } else if (type == 'Mkt') {
+                if (!(("MktPrice" %in% colnames(MktData)) & ("MktVolmn" %in% colnames(MktData)))) stop("No MktPrice or MktVolmn column found, what did you call them?")
+                vwapPrice <- crossprod(MktData[, "MktPrice"], MktData[, "MktVolmn"])/sum(MktData[, "MktVolmn"])
+              }
+              benchPrice <- vwapPrice
+              
+              out <- as.data.frame(cbind(Symbol, c("Buy", "Sell")[side], p_avg, benchPrice), stringsAsFactors = FALSE)
+              colnames(out) <- c('Symbol', 'Side', 'Avg.Exec.Price', paste(benchmark[i], type, sep = '.'))
+            },
+            PWP = {
+              if (is.null(POV)) stop(paste("POV rate needed to compute", benchmark))
+              # if (!("MktVolmn" %in% colnames(MktData))) stop("No MktVolmn column found, what did you call it?")
+              # if (missing(arrTime)) arrTime <- first(index(txns))
+              
+              pwpShares <- rep(tTxnQty/POV, length(txns$Txn.Price))
```

Comment:
See https://github.com/braverock/blotter/pull/93#discussion_r290834253

Reply:
So here as well PWP price needs to be based on market prices. 
Also with respect to the arrival time, as we are assuming all around transactions added to the symbol are the order, thus arrival time would simply be the first transaction time. Hence no need for an explicit new time param here, right?

Reply:
Assuming Arrival Time is equivalent to when the first trade occurred is sufficient for now. As discussed, Arrival Time is when your order hits the order book, which may be minutes or longer before the first trade time, and would impact the market so is a more accurate starting timestamp to begin benchmarking.

I will make code suggestion for PWP shortly. Since multi-line suggestions are not possible in GitHub i will likely just post snippet of PWP code here.

Reply:
Below is my PWP code suggestion. This needs to be tested, but is identical to my suggestion for VWAP except that end.idx is not the index of the final txn used for calculating our VWAP, but the index of the final market txn where cumsum(txns) equates to PWP Shares.

```
            PWP = {
              if (is.null(POV)) stop(paste("POV rate needed to compute", benchmark))
              # if (!("MktVolmn" %in% colnames(MktData))) stop("No MktVolmn column found, what did you call it?")
              # if (missing(arrTime)) arrTime <- first(index(txns))
              
              pwpShares <- tTxnQty/POV
              inds <- which(strftime(index(MktData), format = "%H:%M:%S", tz = "UTC") == strftime(index(txns), format = "%H:%M:%S", tz = "UTC"))
              inds <- sort(inds)
              start.idx <- first(inds)
              # now we need the final timestamp at which cumsum volume equates to PWP Shares...this wont be exact, but we will get close
              # we will use findInterval() for this...i love findInterval() 
              end.idx <- findInterval(pwpShares, cumsum(MktData$MktVolmn[start.idx:nrow(MktData)])) # end.idx is the final timestamp at which cumsum volume equates to PWP Shares
              benchPrice <- crossprod(MktData$MktPrice[start.idx:end.idx], MktData$MktVolmn[start.idx:end.idx])/sum(MktData$MktVolmn[start.idx:end.idx]) # PWP price
              
              out <- as.data.frame(cbind(Symbol, c("Buy", "Sell")[side], p_avg, POV, sum(pwpShares), benchPrice), stringsAsFactors = FALSE)
              colnames(out) <- c('Symbol', 'Side', 'Avg.Exec.Price', 'POV', 'PWP.Shares', 'PWP.Price')
            },
            RPM = {
```

Reply:
Thank you @jaymon0703, your clarifications really helped me demystify some tricky definitions.
There is however something that may be perhaps an unintended behaviour of `findInterval()` in our context. To give an explanatory example, say our `pwpShares` are 14500. Using our testing data and all the rest being the same:

```
> ind <– findInterval(14500, cumsum(ABC[, "MktVolmn"]))
> ind
 [1] 7
```

but at this time

```
> cumsum(ABC[, "MktVolmn"])[ind]
                                  MktVolmn
2019-02-01 09:04:46     4104
```
which is considerably lower than our 14500 shares. In fact, 14500 is much closer to the next observed cumulative market volume, 14882. Results are the same until this last value is reached.

I have tried that function with its different options, but the result was the same. Perhaps I haven't considered one of its aspects and we can indeed use it, please point them out. 

In the meanwhile, I am committing an alternative solution for you to consider. It is still an approximation but we will get closer.

Reply:
The typical order duration someone would want to benchmark will be at least 60mins. findInterval() in my suggestion finds the index at which cumsum(MktData) is closest to PWP Shares. Adding the volume at the next index will take us above PWP Shares. What this means is that we never get to PWP Shares exactly when computing PWP benchmark...but we get close...and for a realistic order with duration 60+mins the impact of the next trade on the PWP Price will be negligible.

In summary, i would not worry about not getting to PWP Shares exactly, this is impossible to do...unless we use the portion of the volume at the next index which makes our cumsum(MktData) == PWP Shares. This will be exact and theoretically more correct, but the benefit over my suggestion should be negligible for typical orders which will be benchmarked in production (ie. 60+mins).

Reply:
Agree on the added complexity needed, but the solution I am about to propose is very simple and perhaps it is worth it.